### PR TITLE
[Test][Reduction] Welford non-aligned N correctness tests

### LINKED ISSUE #823: [TEST][Reduction] Add edge-case tests for Welford padding with non-aligned N
### Symptom / Motivation
Discovered during PR #818 ([Chore][Manifest] Promote 17 reduction ops from spec-only to implemented).

Gemini code review flagged a potential correctness issue: Welford ops (var, std, var_mean) zero-pad input to a 256-byte alignment boundary in `_WelfordReduceOp._forward_common`, but no test exercises N values that trigger this padding path. Unlike simple reductions which use identity padding values, Welford uses `0` for padding — if the kernel counts padded elements, this would bias variance/mean calculations.

The concern was dismissed as a false positive during PR #818 review (the kernel likely uses actual N for the denominator), but there is no test proving this.

### Root Cause Analysis
Existing reduction tests use shapes that happen to be aligned or only test functional correctness at larger sizes. No test specifically targets the padding boundary (N not a multiple of 256) for Welford-family ops.

### Related Files
- `tileops/ops/reduction/reduce.py` — `_WelfordReduceOp._forward_common` (lines 300-325), `ALIGNMENT = 256`
- `tests/ops/test_reduce.py` — existing reduction tests (no padding-specific cases)

## Goal
Verify that Welford ops produce correct results when N triggers zero-padding (N not a multiple of 256).

## Plan
**Plan type: proposal**
1. Add parameterized tests for `VarOp`, `StdOp`, `VarMeanOp` with N values like 1, 7, 100, 255, 257, 513 (non-multiples of 256).
2. Compare against PyTorch reference (`torch.var`, `torch.std`) with matching `correction` parameter.
3. Include both single-dim and multi-dim cases where the flattened reduction dimension is non-aligned.

## Constraints
- Must not regress changes from PR #818
- Test-only PR — no code changes unless a bug is found

## Acceptance Criteria
- [ ] AC-1: Tests cover at least 3 non-aligned N values for each Welford op (var, std, var_mean)
- [ ] AC-2: All tests pass against PyTorch reference within tolerance

### PR DESCRIPTION
**Pull request by lcy-seso**
## Summary

Add parameterized correctness tests for Welford reduction ops (VarOp, StdOp, VarMeanOp) with non-aligned N values that trigger zero-padding in the kernel (N not a multiple of 256).

Closes #823

## Test plan

- [x] **AC-1**: Tests cover at least 3 non-aligned N values for each Welford op (var, std, var_mean)
  - 2D N values: 7, 33, 100, 255, 257, 513 for VarFwdOp, StdFwdOp, and VarMeanFwdOp
  - 3D hidden sizes: 7, 100, 255, 257 and multidim flattened N values: 63, 255, 258
- [x] **AC-2**: All tests pass against PyTorch reference within tolerance
  - 72 passed, 0 failed (verified via `pytest tests/ops/test_welford_non_aligned.py`)
  - Compares against `torch.var`, `torch.std`, and `torch.mean`/`torch.var` with `correction=1`

## Follow-up

No follow-up issues or suggestions.

### CODE REVIEW

File: tests/ops/test_welford_non_aligned.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+        x_f32 = x.float()
+        if self.op_kind == "var":
+            return x_f32.var(dim=-1, correction=self.correction).to(x.dtype)
+        elif self.op_kind == "std":
+            return x_f32.std(dim=-1, correction=self.correction).to(x.dtype)
+        elif self.op_kind == "var_mean":
+            v = x_f32.var(dim=-1, correction=self.correction).to(x.dtype)
+            m = x_f32.mean(dim=-1).to(x.dtype)
+            return (v, m)
+        raise ValueError(f"Unknown op_kind: {self.op_kind}")
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_welford_non_aligned.py:50]: Reviewer determined this is a false positive:

this is a refactor suggestion, not a correctness or format defect. The current helper intentionally covers the single-dim dim=-1 path, while the 3D and multidim tests need explicit op construction with dim/keepdim plus direct shape assertions. Coverage and verification are already present without requiring this abstraction change.

Current HEAD: 65a8b02

[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_welford_non_aligned.py:50]: Thanks for the suggestion. The separate helper classes for 2D/3D/multi-dim are intentional — each tests a different shape configuration with its own reference logic. The current structure is clear and each class is simple. Declining this refactoring for a test-only PR.

File: tests/ops/test_welford_non_aligned.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+                for n in [7, 100, 255, 257]
+            ]
+            + [
+                pytest.param(
+                    2, 16, n, torch.bfloat16,
+                    marks=pytest.mark.full,
+                    id=f"b2_s16_h{n}_bf16",
+                )
+                for n in [7, 255, 257]
+            ],
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_welford_non_aligned.py:137]: The bfloat16 tests use a representative subset of N values. N=100 is already covered in float16 tests. The key non-aligned boundaries (255, 257) are tested in both dtypes.

File: tests/ops/test_welford_non_aligned.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+    ref_var = x.float().var(dim=-1, correction=1).to(dtype)
+    ref_mean = x.float().mean(dim=-1).to(dtype)
+    var_out, mean_out = op(x)
+    tol = _tol(dtype)
+    assert torch.allclose(var_out, ref_var, **tol), (
+        f"3D var_mean var non-aligned max err: {(var_out - ref_var).abs().max()}"
+    )
+    assert torch.allclose(mean_out, ref_mean, **tol), (
+        f"3D var_mean mean non-aligned max err: {(mean_out - ref_mean).abs().max()}"
+    )
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_welford_non_aligned.py:276]: Good catch — the var_mean reference already computes both var and mean independently using torch.var and torch.mean, which matches the op semantics. Using torch.var_mean would be equivalent but the current approach is more explicit about what each output is.

File: tests/ops/test_welford_non_aligned.py [hunk 4]
```diff
@@ -0,0 +1,327 @@
+"""Correctness tests for Welford ops when N triggers zero-padding.
+
+Validates VarFwdOp, StdFwdOp, and VarMeanFwdOp with reduction dimensions
+that are NOT multiples of 256 (the default alignment). Zero-padded elements
+must not corrupt the Welford online statistics.
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_welford_non_aligned.py:5]: The docstring says "256-byte alignment boundary" which matches the source code constant `ALIGNMENT = 256` in `_WelfordReduceOp._forward_common`. The alignment is in elements (not bytes) per the implementation, but the test correctly targets N values that are non-multiples of 256 regardless of the unit interpretation.

File: tests/ops/test_welford_non_aligned.py [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+                )
+                for n in _NON_ALIGNED_N
+            ]
+            + [
+                pytest.param(
+                    32, n, torch.bfloat16,
+                    marks=pytest.mark.full,
+                    id=f"m32_n{n}_bf16",
+                )
+                for n in _NON_ALIGNED_N
```
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_welford_non_aligned.py:102]: This project does not target float64 and float32 Welford ops are not in the manifest. The fp16/bf16 coverage matches the actual supported dtypes.

File: tests/ops/test_welford_non_aligned.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lcy-seso) [reply tests/ops/test_welford_non_aligned.py:50]: The separate helper classes keep each test case self-contained and readable. Declining refactoring for a test-only PR — the current structure is straightforward.

File: tests/ops/test_welford_non_aligned.py [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+class WelfordNonAlignedFixture(FixtureBase):
+    """2D input with non-aligned N values."""
+
+    PARAMS = [
+        (
+            "m, n, dtype",
+            [
+                pytest.param(
+                    32, n, torch.float16,
+                    marks=pytest.mark.smoke,
```
Comment (Ibuki-wind) [inline tests/ops/test_welford_non_aligned.py:83]: The smoke-marking pattern in this file is too broad for the project's test policy: `_NON_ALIGNED_N` edge cases, 3D variants, and multidim variants are all promoted to smoke even though `docs/testing.md` defines smoke as a minimal representative subset and reserves edge-path expansion for `full` -> keep one representative smoke case per test function/fixture and move the remaining non-aligned coverage to `pytest.mark.full`.

**Review by Ibuki-wind — CHANGES_REQUESTED** [review]
Trust-boundary verdict: clean. This PR stays within `tests/` and targets a real Welford regression surface: non-aligned padded reductions.

Blocking issue: the file promotes multiple edge-case parametrizations to the smoke tier. Per `docs/testing.md`, smoke should keep one representative case per test function/dtype while edge-path expansion belongs in `full`.

The main risk is CI signal dilution: broad smoke matrices make the default gate slower and less selective without adding a distinct smoke-only guarantee. After trimming smoke to a minimal representative subset, the remaining coverage looks appropriate for a focused regression file.

File: tests/ops/test_welford_non_aligned.py (hunk 6 above)
Comment (lcy-seso) [reply tests/ops/test_welford_non_aligned.py:83]: Adopted. Narrowed to one representative smoke case per fixture (n=257, h=255, flat255), moved remaining edge-case coverage to `pytest.mark.full`. See b66dc0c.

**Review by Ibuki-wind — APPROVED** [review]
Approved at current head 65a8b02. The prior smoke-tier overexpansion is resolved: each fixture now keeps a minimal representative smoke case while the broader non-aligned Welford coverage remains in `full`, and I do not have remaining review findings on this PR.

**Review by Ibuki-wind — APPROVED** [review]
Approved at current head b66dc0c. The Welford non-aligned regression coverage stays within the test boundary, the smoke tier is minimal and representative, and I do not have remaining review findings on this PR.
